### PR TITLE
Client library updating and client library feature awareness

### PR DESCRIPTION
**Pull request by aptkingston**
## Description
This PR adds the ability to update the client library an app uses from within the builder. You will also be able to revert to the previous version of your client library after updating, which is useful in case of accidental updates or issues with new versions. It's also a nice fail safe incase we ship a bug as it gives users an escape hatch to go back to the previous version.

This also adds a feature definition section of the manifest file of the client library, which the builder now uses to know what features are supported. This means the builder is aware of what features it should show that are supported by the app being developed. A good example of this is the new spectrum theme picker - the builder will now only show the theme picker when you're using a new enough client library to support it.

As part of this work I've refactored how the server fetches the required client library files, and removed the dependency on pulling down tarballs from NPM. When updating, the server will instead serve `manifest.json` and `budibase-client.js` from the local filesystem, which will always be the latest version. This might also solve https://github.com/Budibase/budibase/issues/1866, but it depends on what way the deps are installed in the server image. I'll need to check this whole feature actually, because it depends on the server being bundled with the latest client library which might only happen in dev due to lerna symlinking.

## Screenshots
Updating and reverting is controlled by a new icon in the top-right of the builder. This icon is blue when an update is available.
image

This opens a modal explaining that there is an update available.
image

Opening the modal when there is no update available will explain that they are on the latest version. It will also allow reverting if there is a version to revert to.
image

### CODE REVIEW

File: packages/server/src/utilities/fileSystem/clientLibrary.js [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    // Load the bundled version in prod
+    manifest = resolve(TOP_LEVEL_PATH, "client", "manifest.json")
+    client = resolve(TOP_LEVEL_PATH, "client", "budibase-client.js")
+
+    console.log(manifest)
+    console.log(client)
+  }
+
+  // Upload latest component manifest
+  await streamUpload(
```
Comment (shogunpurple) [inline packages/server/src/utilities/fileSystem/clientLibrary.js:108]: NAB: Probably a micro-optimisation but these 2 `streamUpload` calls will run in series. If we use `Promise.all` we can kick them off concurrently

File: packages/server/src/utilities/fileSystem/clientLibrary.js [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+  )
+
+  // Copy backup client lib to tmp
+  tmpClientPath = await retrieveToTmp(
+    ObjectStoreBuckets.APPS,
+    join(appId, "budibase-client.js.bak")
+  )
+
+  // Upload manifest backup
+  await upload({
```
Comment (shogunpurple) [inline packages/server/src/utilities/fileSystem/clientLibrary.js:151]: See above commment

**Review by shogunpurple — APPROVED** [review]
LGTM! Just a few optional comments.

I really like the way this works, in that we only store one revertable version. Full history is a nightmare to maintain. Great job 🚀

File: packages/server/src/api/controllers/application.js [hunk 3]
```diff
@@ -269,6 +301,25 @@ exports.delete = async function (ctx) {
...
+
+  const newAppPackage = { ...application, ...appPackage, url }
+  if (appPackage._rev !== application._rev) {
+    newAppPackage._rev = application._rev
+  }
+
+  // the locked by property is attached by server but generated from
+  // Redis, shouldn't ever store it
+  if (newAppPackage.lockedBy) {
+    delete newAppPackage.lockedBy
```
[the remark this answers is not in the document]
Comment (aptkingston) [reply packages/server/src/api/controllers/application.js:317]: Yep good point. I didn't actually write this, I just moved it as part of the refactoring but may as well update it.

File: packages/server/src/utilities/fileSystem/clientLibrary.js (hunk 1 above)
Comment (aptkingston) [reply packages/server/src/utilities/fileSystem/clientLibrary.js:108]: Good call. Every few milliseconds helps haha.

**Comment by aptkingston** [comment]
> 
> 
> LGTM! Just a few optional comments.
> 
> I really like the way this works, in that we only store one revertable version. Full history is a nightmare to maintain. Great job 🚀

Cheers. I think maintaining just one revertable verstion makes sense. Like you say it massively reduces the maintenance, but hopefully no one would ever need to revert back more than one version anyway. I mostly see it as an escape hatch in case you either accidentally upgrade, or upgrade and experience issues or a bug and want to downgrade temporarily.

**Pull request merged**